File: src/main/java/com/kgaft/VulkanLib/Device/SwapChain.java
package com.kgaft.VulkanLib.Device;

import java.lang.instrument.IllegalClassFormatException;
import java.nio.IntBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;

import com.kgaft.VulkanLib.Utils.DestroyableObject;
import com.kgaft.VulkanLib.Utils.LwjglObject;
import com.kgaft.VulkanLib.Utils.SeriesObject;
import com.kgaft.VulkanLib.Utils.VkErrorException;
import org.lwjgl.vulkan.*;

import com.kgaft.VulkanLib.Device.Image.Image;
import com.kgaft.VulkanLib.Device.Image.ImageView;
import com.kgaft.VulkanLib.Device.LogicalDevice.LogicalDevice;

import static org.lwjgl.vulkan.VK13.*;
import static org.lwjgl.vulkan.KHRSurface.*;
import static org.lwjgl.vulkan.KHRSwapchain.*;

class SwapChainSupportDetails {
    public LwjglObject<VkSurfaceCapabilitiesKHR> capabilities = new LwjglObject<>(VkSurfaceCapabilitiesKHR.class);
    public LwjglObject<VkSurfaceFormatKHR.Buffer> formats;
    public int[] presentModes;

    SwapChainSupportDetails() throws IllegalClassFormatException {
    }
}

public class SwapChain extends DestroyableObject {
    private static LwjglObject<VkSurfaceFormatKHR.Buffer> formatsBuffer;
    private static VkSurfaceFormatKHR format;
    private static SeriesObject<VkSwapchainCreateInfoKHR> createInfos;
    static {
        try {
            createInfos = new SeriesObject<>(VkSwapchainCreateInfoKHR.class);
        } catch (IllegalClassFormatException e) {
            e.printStackTrace();
        }
    }

    public static VkSurfaceFormatKHR getFormat() {
        return format;
    }

    private LogicalDevice device;
    private long swapchainKhr = 0;
    private long surface;
    private int presentMode;
    private LwjglObject<VkExtent2D> extent;
    private long[] baseImages;
    private List<Image> swapchainImages = new ArrayList<>();
    private List<ImageView> swapchainImageViews = new ArrayList<>();
    private int width;
    private int height;
    private boolean enableFrameLock = false;

    public SwapChain(LogicalDevice device, long surface,
                     int width, int height, boolean enableFrameLock) throws IllegalClassFormatException, VkErrorException {
        this.device = device;
        this.surface = surface;
        this.width = width;
        this.height = height;
        this.enableFrameLock = enableFrameLock;
        createSwapChain(width, height, enableFrameLock);
    }

    public void recreate(int width, int height) throws IllegalClassFormatException, VkErrorException {

        this.width = width;
        this.height = height;
        cleanUpImages();
        createSwapChain(width, height, enableFrameLock);
    }

    public void recreate(int width, int height, boolean refreshRateLock) throws IllegalClassFormatException, VkErrorException {

        enableFrameLock = refreshRateLock;
        this.width = width;
        this.height = height;
        cleanUpImages();
        createSwapChain(width, height, refreshRateLock);
    }

    public List<Image> getSwapchainImages() {
        return swapchainImages;
    }

    public List<ImageView> getSwapchainImageViews() {
        return swapchainImageViews;
    }

    public long getSwapchainKhr() {
        return swapchainKhr;
    }

    private void createSwapChain(int width, int height, boolean enableFrameLock) throws IllegalClassFormatException, VkErrorException {
        SwapChainSupportDetails support = new SwapChainSupportDetails();

        gatherSwapChainSupportDetails(support);
        formatsBuffer = support.formats;
        format = chooseSurfaceFormat(support.formats);
        presentMode = choosePresentMode(support.presentModes, enableFrameLock);
        extent = chooseSwapchainExtent(width, height, support.capabilities);
        int imageCount = support.capabilities.get().minImageCount() + 1;
        if (support.capabilities.get().maxImageCount() > 0 &&
                imageCount > support.capabilities.get().maxImageCount()) {
            imageCount = support.capabilities.get().maxImageCount();
        }
        VkSwapchainCreateInfoKHR createInfo = createInfos.acquireObject();
        createInfo.sType$Default();
        createInfo.surface(surface);
        createInfo.minImageCount(imageCount);
        createInfo.imageFormat(format.format());
        createInfo.imageColorSpace(format.colorSpace());
        createInfo.imageExtent(extent.get());
        createInfo.imageArrayLayers(1);
        createInfo.imageUsage(VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT);
        HashSet<Integer> baseQIndices = new HashSet<Integer>();
        baseQIndices.add(device.getQueueByType(VK_QUEUE_GRAPHICS_BIT).getIndex());
        baseQIndices.add(device.getPresentQueue().getIndex());


        IntBuffer queueIndices = IntBuffer.allocate(baseQIndices.size());
        baseQIndices.forEach(queueIndices::put);
        queueIndices.rewind();
        if (queueIndices.capacity() > 1) {
            createInfo.imageSharingMode(VK_SHARING_MODE_CONCURRENT);
            createInfo.pQueueFamilyIndices(queueIndices);
        } else {
            createInfo.imageSharingMode(VK_SHARING_MODE_EXCLUSIVE);
        }
        createInfo.preTransform(support.capabilities.get().currentTransform());
        createInfo.compositeAlpha(VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR);
        createInfo.presentMode(presentMode);
        createInfo.clipped(true);
        createInfo.oldSwapchain(swapchainKhr);
        long[] preSwap = new long[1];
        VkErrorException.checkVkStatus("Failed to create swapchain", vkCreateSwapchainKHR(device.getDevice(), createInfo, null, preSwap));
        this.swapchainKhr = preSwap[0];
        int[] swapImagesC = new int[1];
        createInfos.releaseObjectInstance(createInfo);
        vkGetSwapchainImagesKHR(device.getDevice(), swapchainKhr, swapImagesC, null);
        baseImages = new long[swapImagesC[0]];
        vkGetSwapchainImagesKHR(device.getDevice(), swapchainKhr, swapImagesC, baseImages);
        
        Arrays.stream(baseImages).forEach(element -> {
            try {
                Image img = new Image(device, element);
                swapchainImages.add(img);
                LwjglObject<VkImageViewCreateInfo> viewCreateInfo = new LwjglObject<>(VkImageViewCreateInfo.class);
                viewCreateInfo.get().image(element);
                viewCreateInfo.get().viewType(VK_IMAGE_VIEW_TYPE_2D);
                viewCreateInfo.get().format(format.format());
                viewCreateInfo.get().components().r(VK_COMPONENT_SWIZZLE_IDENTITY);
                viewCreateInfo.get().components().g(VK_COMPONENT_SWIZZLE_IDENTITY);
                viewCreateInfo.get().components().b(VK_COMPONENT_SWIZZLE_IDENTITY);
                viewCreateInfo.get().components().a(VK_COMPONENT_SWIZZLE_IDENTITY);
                viewCreateInfo.get().subresourceRange().aspectMask(VK_IMAGE_ASPECT_COLOR_BIT);
                viewCreateInfo.get().subresourceRange().baseMipLevel(0);
                viewCreateInfo.get().subresourceRange().levelCount(1);
                viewCreateInfo.get().subresourceRange().baseArrayLayer(0);
                viewCreateInfo.get().subresourceRange().layerCount(1);
                swapchainImageViews.add(img.createImageView(viewCreateInfo));

            } catch (IllegalClassFormatException e) {
                throw new RuntimeException(e);
            } catch (VkErrorException e) {
                throw new RuntimeException(e);
            }

        });

    }

    private void gatherSwapChainSupportDetails(SwapChainSupportDetails output) throws VkErrorException, IllegalClassFormatException {
        int res = vkGetPhysicalDeviceSurfaceCapabilitiesKHR(device.getBase().getBase(), surface, output.capabilities.get());

        VkErrorException.checkVkStatus("Failed to gather surface capabilities", res);
        int[] formatCount = new int[1];
        res = vkGetPhysicalDeviceSurfaceFormatsKHR(device.getBase().getBase(), surface, formatCount, null);
        VkErrorException.checkVkStatus("Failed to gather surface formats", res);
        output.formats = new LwjglObject<>(VkSurfaceFormatKHR.class, VkSurfaceFormatKHR.Buffer.class, formatCount[0]);
        res = vkGetPhysicalDeviceSurfaceFormatsKHR(device.getBase().getBase(), surface, formatCount, output.formats.get());
        VkErrorException.checkVkStatus("Failed to gather surface formats", res);
        res = vkGetPhysicalDeviceSurfacePresentModesKHR(device.getBase().getBase(), surface, formatCount, null);
        VkErrorException.checkVkStatus("Failed to gather surface present modes", res);
        output.presentModes = new int[formatCount[0]];
        res = vkGetPhysicalDeviceSurfacePresentModesKHR(device.getBase().getBase(), surface, formatCount, output.presentModes);
        VkErrorException.checkVkStatus("Failed to gather surface present modes", res);
    }

    private VkSurfaceFormatKHR
    chooseSurfaceFormat(LwjglObject<VkSurfaceFormatKHR.Buffer> formats) {
        int[] preferredFormats = new int[]{VK_FORMAT_R64G64B64A64_SFLOAT, VK_FORMAT_R32G32B32A32_SFLOAT, VK_FORMAT_R16G16B16A16_SFLOAT, VK_FORMAT_R16G16B16A16_SNORM,
                VK_FORMAT_R16G16B16A16_UNORM, VK_FORMAT_R16G16B16A16_USCALED, VK_FORMAT_R16G16B16A16_SSCALED, VK_FORMAT_B16G16R16G16_422_UNORM, VK_FORMAT_R8G8B8A8_SRGB, VK_FORMAT_B8G8R8A8_SRGB};
        VkSurfaceFormatKHR selectedFormat = null;
        for (VkSurfaceFormatKHR item : formats.get()) {
            for (int citem : preferredFormats) {
                if (item.format() == citem) {
                    if (selectedFormat != null) {
                        if (item.format() > selectedFormat.format() || (item.colorSpace() >= selectedFormat.colorSpace() && item.format() >= selectedFormat.format())) {
                            selectedFormat = item;
                        }
                    } else {
                        selectedFormat = item;
                    }

                }
            }
        }
        if (selectedFormat != null) {
            return selectedFormat;
        }
        formats.get().rewind();
        return formats.get().get();
    }

    private int
    choosePresentMode(int[] presentModes, boolean enableFrameLock) {
        for (int presentMode : presentModes) {
            if (presentMode == (enableFrameLock ? VK_PRESENT_MODE_MAILBOX_KHR : VK_PRESENT_MODE_IMMEDIATE_KHR)) {
                return presentMode;
            }
        }

        return VK_PRESENT_MODE_FIFO_KHR;
    }

    private LwjglObject<VkExtent2D> chooseSwapchainExtent(int width, int height,
                                                          LwjglObject<VkSurfaceCapabilitiesKHR> capabilities) throws IllegalClassFormatException {

        if (capabilities.get().currentExtent().width() != Integer.MAX_VALUE) {
            return new LwjglObject<>(capabilities.get().currentExtent(), VkExtent2D.class);
        } else {
            LwjglObject<VkExtent2D> res = new LwjglObject<>(VkExtent2D.class);
            res.get().width(width);
            res.get().height(height);
            extent.get().width(
                    Math.min(capabilities.get().maxImageExtent().width(),
                            Math.max(capabilities.get().minImageExtent().width(), extent.get().width())));

            extent.get().height(
                    Math.min(capabilities.get().maxImageExtent().height(),
                            Math.max(capabilities.get().minImageExtent().height(), extent.get().height())));

            return extent;
        }
    }

    private void cleanUpImages(){
        swapchainImageViews.forEach(ImageView::destroy);
        swapchainImageViews.clear();
        swapchainImages.clear();
    }
    @Override
    public void destroy() {
        destroyed = true;
        cleanUpImages();
        vkDestroySwapchainKHR(device.getDevice(), swapchainKhr, null);
    }
}

File: src/device/swapchain.rs
use crate::device::image::image::VlImage;
use crate::device::image::image_view::VlImageView;
use crate::device::logical_device::logical_device::VlLogicalDevice;
use crate::instance::instance::VlInstance;
use ash::khr::swapchain;
use ash::vk;
use std::collections::HashSet;

#[derive(Default)]
struct SwapChainDetails {
    capabilities: vk::SurfaceCapabilitiesKHR,
    formats: Vec<vk::SurfaceFormatKHR>,
    present_modes: Vec<vk::PresentModeKHR>,
}
/**
    VlSwapChain is not supported shairing, like other classes,
    if you need to share it between threads, use Arc Mutex,
    images from swapchain not shared as well, despite that other images are shared compatible via clone
*/
pub struct VlSwapChain {
    format: vk::SurfaceFormatKHR,
    present_mode: vk::PresentModeKHR,
    extent: vk::Extent2D,
    images: Vec<VlImage>,
    image_views: Vec<VlImageView>,
    swap_chain_loader: swapchain::Device,
    device: VlLogicalDevice,
    instance: VlInstance,
    swap_chain: vk::SwapchainKHR,
    surface: vk::SurfaceKHR,
    frame_lock: bool
}

impl Drop for VlSwapChain {
    fn drop(&mut self) {
        self.destroy();
    }
}

unsafe impl Send for VlSwapChain {

}

unsafe impl Sync for VlSwapChain {

}

impl VlSwapChain {
    pub fn new(
        device: VlLogicalDevice,
        instance: VlInstance,
        surface: vk::SurfaceKHR,
        width: u32,
        height: u32,
        frame_lock: bool,
    ) -> Self {
        let swap_chain_loader =
            swapchain::Device::new(instance.get_instance_r(), device.device_r());
        let mut pre_res = Self {
            format: Default::default(),
            device,
            instance,
            swap_chain_loader,
            surface,
            present_mode: Default::default(),
            extent: Default::default(),
            images: vec![],
            image_views: vec![],
            swap_chain: Default::default(),
            frame_lock
        };
        pre_res.create_swap_chain(width, height, frame_lock, surface);
        pre_res
    }

    fn create_swap_chain(
        &mut self,
        width: u32,
        height: u32,
        frame_lock: bool,
        surface: vk::SurfaceKHR,
    ) {
        let mut support = SwapChainDetails::default();
        Self::gather_swap_chain_info(&self.instance, &self.device, surface, &mut support);
        self.format = Self::choose_surface_format(support.formats.as_slice());
        self.present_mode = Self::choose_present_mode(support.present_modes.as_slice(), frame_lock);
        self.extent = Self::choose_swapchain_extent(width, height, &support.capabilities);
        let mut image_count = support.capabilities.min_image_count + 1;
        if support.capabilities.max_image_count > 0
            && image_count > support.capabilities.max_image_count
        {
            image_count = support.capabilities.max_image_count;
        }

        let mut queue_indices: Vec<u32> = vec![
            self.device
                .find_queue_by_type_r(vk::QueueFlags::GRAPHICS)
                .unwrap()
                .index(),
            self.device.find_present_queue_r().unwrap().index(),
        ];
        let unique_indices: HashSet<u32> = queue_indices.drain(..).collect();
        let mut queue_indices: Vec<u32> = unique_indices.into_iter().collect();

        // Ensure stable ordering (important for safety across drivers)
        queue_indices.sort_unstable();

        // Default create info
        let mut create_info = vk::SwapchainCreateInfoKHR::default()
            .surface(surface)
            .min_image_count(image_count)
            .image_format(self.format.format)
            .image_color_space(self.format.color_space)
            .image_extent(self.extent)
            .image_array_layers(1)
            .image_usage(vk::ImageUsageFlags::COLOR_ATTACHMENT)
            .pre_transform(support.capabilities.current_transform)
            .composite_alpha(vk::CompositeAlphaFlagsKHR::OPAQUE)
            .present_mode(self.present_mode)
            .clipped(true)
            .old_swapchain(vk::SwapchainKHR::null());

        if queue_indices.len() > 1 {
            create_info = create_info
                .image_sharing_mode(vk::SharingMode::CONCURRENT)
                .queue_family_indices(&queue_indices);
        } else {
            create_info = create_info.image_sharing_mode(vk::SharingMode::EXCLUSIVE);
        }
        self.swap_chain = unsafe { self.swap_chain_loader.create_swapchain(&create_info, None) }
            .expect("Failed to create swap chain");
        let mut images = unsafe {
            self.swap_chain_loader
                .get_swapchain_images(self.swap_chain.clone())
                .expect("Failed to get swap chain images")
        };

        while let Some(image) = images.pop() {
            let mut image = VlImage::from_image(self.device.device(), image);
            let view_create_info = vk::ImageViewCreateInfo {
                image: image.image(),
                view_type: vk::ImageViewType::TYPE_2D,
                format: self.format.format,
                components: vk::ComponentMapping {
                    r: vk::ComponentSwizzle::IDENTITY,
                    g: vk::ComponentSwizzle::IDENTITY,
                    b: vk::ComponentSwizzle::IDENTITY,
                    a: vk::ComponentSwizzle::IDENTITY,
                },
                subresource_range: vk::ImageSubresourceRange {
                    aspect_mask: vk::ImageAspectFlags::COLOR,
                    base_mip_level: 0,
                    level_count: 1,
                    base_array_layer: 0,
                    layer_count: 1,
                },
                ..Default::default()
            };
            let view = image.create_image_view(view_create_info);
            self.images.push(image);
            self.image_views.push(view);
        }
    }

    pub fn recreate_swap_chain(&mut self, width: u32, height: u32){
        self.destroy();
        self.create_swap_chain(width, height, self.frame_lock, self.surface);
    }
    
    



    pub fn recreate_swap_chain_framelock(&mut self, width: u32, height: u32, frame_lock: bool){
        self.frame_lock = frame_lock;
        self.destroy();
        self.create_swap_chain(width, height, self.frame_lock, self.surface);
    }



    fn cleanup_images(&mut self) {
        while let Some(view) = self.image_views.pop() {
            drop(view)
        }
        self.image_views.clear();
        self.images.clear();
    }

    fn destroy(&mut self) {
        self.cleanup_images();
        unsafe {
            self.device.device().device_wait_idle().unwrap();
            self.swap_chain_loader
                .destroy_swapchain(self.swap_chain, None);
        }
    }

    fn gather_swap_chain_info(
        instance: &VlInstance,
        device: &VlLogicalDevice,
        surface: vk::SurfaceKHR,
        res: &mut SwapChainDetails,
    ) {
        let capab = unsafe {
            instance
                .surface_loader()
                .expect("Surface was not loaded")
                .get_physical_device_surface_capabilities(device.base_device().inner(), surface)
        }
        .expect("Failed to get surface capabilities");
        let formats = unsafe {
            instance
                .surface_loader()
                .unwrap()
                .get_physical_device_surface_formats(device.base_device().inner(), surface)
        }
        .expect("Failed to get surface formats");
        let present_modes = unsafe {
            instance
                .surface_loader()
                .unwrap()
                .get_physical_device_surface_present_modes(device.base_device().inner(), surface)
                .expect("Failed to get present modes")
        };
        res.present_modes = present_modes;
        res.capabilities = capab;
        res.formats = formats;
    }



    fn choose_surface_format(formats: &[vk::SurfaceFormatKHR]) -> vk::SurfaceFormatKHR {
        // Expanded preferred formats list (HDR → High → Basic)
        let preferred_formats = [
            // High precision float formats
            vk::Format::R64G64B64A64_SFLOAT,
            vk::Format::R32G32B32A32_SFLOAT,
            vk::Format::R32G32B32A32_SFLOAT,
            // HDR10 / HDR formats
            vk::Format::A2B10G10R10_UNORM_PACK32,
            vk::Format::A2R10G10B10_UNORM_PACK32,
            vk::Format::R16G16B16A16_SFLOAT,
            vk::Format::R16G16B16A16_UNORM,
            vk::Format::R16G16B16A16_SNORM,
            // sRGB formats (most common for monitors)
            vk::Format::B8G8R8A8_SRGB,
            vk::Format::R8G8B8A8_SRGB,
            // Basic UNORM/linear
            vk::Format::B8G8R8A8_UNORM,
            vk::Format::R8G8B8A8_UNORM,
        ];

        let mut selected_format = vk::SurfaceFormatKHR {
            format: vk::Format::UNDEFINED,
            color_space: vk::ColorSpaceKHR::SRGB_NONLINEAR,
        };

        for item in formats {
            if preferred_formats.contains(&item.format) {
                if item.format > selected_format.format
                    || (item.color_space >= selected_format.color_space
                        && item.format >= selected_format.format)
                {
                    selected_format = *item;
                }
            }
        }

        if selected_format.format != vk::Format::UNDEFINED {
            selected_format
        } else {
            // fallback: just pick the first format (guaranteed to exist)
            formats[0]
        }
    }

    /// Choose the present mode depending on frame lock.
    /// Mailbox → Immediate → Fifo fallback
    fn choose_present_mode(
        present_modes: &[vk::PresentModeKHR],
        enable_frame_lock: bool,
    ) -> vk::PresentModeKHR {
        let preferred = if enable_frame_lock {
            vk::PresentModeKHR::MAILBOX
        } else {
            vk::PresentModeKHR::IMMEDIATE
        };

        present_modes
            .iter()
            .copied()
            .find(|&m| m == preferred)
            .unwrap_or(vk::PresentModeKHR::FIFO) // guaranteed available
    }

    /// Choose the swapchain extent based on surface capabilities and desired size.
    /// Clamps to min/max range when current_extent is undefined.
    fn choose_swapchain_extent(
        width: u32,
        height: u32,
        capabilities: &vk::SurfaceCapabilitiesKHR,
    ) -> vk::Extent2D {
        if capabilities.current_extent.width != u32::MAX {
            capabilities.current_extent
        } else {
            let clamped_width = capabilities
                .min_image_extent
                .width
                .max(width)
                .min(capabilities.max_image_extent.width);

            let clamped_height = capabilities
                .min_image_extent
                .height
                .max(height)
                .min(capabilities.max_image_extent.height);

            vk::Extent2D {
                width: clamped_width,
                height: clamped_height,
            }
        }
    }

    pub fn swap_chain_loader(&self) -> &swapchain::Device {
        &self.swap_chain_loader
    }

    pub fn swap_chain(&self) -> vk::SwapchainKHR {
        self.swap_chain
    }

    pub fn format(&self) -> vk::SurfaceFormatKHR {
        self.format
    }

    pub fn images(&self) -> &Vec<VlImage> {
        &self.images
    }

    pub fn image_views(&self) -> &Vec<VlImageView> {
        &self.image_views
    }

    pub fn extent(&self) -> vk::Extent2D {
        self.extent
    }
}
